### PR TITLE
[RFC] refactor/single-include: terminal.h, regexp_defs.h, syntax_defs.h, undo_defs.h, undo.h

### PR DESCRIPTION
**Pull request by relnod**
ref #5321

### CODE REVIEW

File: src/nvim/regexp_defs.h [hunk 1]
```diff
@@ -41,18 +43,38 @@
 #define NFA_ENGINE          2
 
 typedef struct regengine regengine_T;
+typedef struct regprog regprog_T;
+typedef struct reg_extmatch reg_extmatch_T;
```
Comment (justinmk) [inline src/nvim/regexp_defs.h]: Might as well give this treatment to regmmatch_T too.

Also the linter will want the `/*`-style comments changed to `//`-style.

File: src/nvim/CMakeLists.txt [hunk 2]
```diff
@@ -526,11 +526,6 @@ endfunction()
 
 set(NO_SINGLE_CHECK_HEADERS
   os/win_defs.h
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt]: I don't think this one can be checked at all, so @ZyX-I I think we can close https://github.com/neovim/neovim/issues/5321 ?

File: src/nvim/regexp_defs.h (hunk 1 above)
Comment (relnod) [inline src/nvim/regexp_defs.h]: The whole definition of regmmatch_T is needed before the include of "buffer_defs.h". Should i still split it into a typedef struct and struct?

File: src/nvim/CMakeLists.txt (hunk 2 above)
Comment (ZyX-I) [inline src/nvim/CMakeLists.txt]: This may be checked on Windows, AFAIR I was suggesting to add this to NO_SINGLE_CHECK_HEADERS depending on the platform condition.

File: src/nvim/regexp_defs.h (hunk 1 above)
Comment (justinmk) [inline src/nvim/regexp_defs.h]: Oh, right.

Ideally, circular dependencies would be resolved by splitting things up. See @ZyX-I advice in https://github.com/neovim/neovim/issues/5321#issuecomment-292217974 and https://github.com/neovim/neovim/issues/6371

tl;dr: What you have posted _works_, so don't change it. But if you are inclined, see if a new header can be introduced to avoid this fragile ordering.

**Comment by relnod** [comment]
@justinmk I tried splitting regexp_defs.h to avoid the circular dependency. Though, I'm not sure if this is the correct approach.

File: src/nvim/regexp_defs.h (hunk 1 above)
Comment (justinmk) [reply src/nvim/regexp_defs.h]: @relnod Is this ordering still needed, now that buffer_defs.h includes the new regexp_2_defs.h ?

I see the last commit is marked WIP. Do you want to finish it, or maybe break it into a new PR, so that we can merge the previous changes?

**Comment by relnod** [comment]
@justinmk I removed the last commit, where I tried to split the regexp_defs.h. I might try it again in a seperate pr. Also rebased on master.
